applications/Chat/evaluate/gpt_evaluate.py

```diff
@@ -0,0 +1,525 @@
+
+    if os.path.exists(os.path.join(save_path, "battle_results.json")):
+        results = jload(os.path.join(save_path, "battle_results.json"))
+    else:
+        results = {}
+
+    results[prefix] = {
+        "model": [name1, name2],
+        "better": better_count,
+        "worse": worse_count,
+        "tie": tie_count,
+        "win_rate": better_count / (len(evaluations) - invalid_count),
+        "score": [
+            ans1_score / (len(evaluations) - invalid_count),
+            ans2_score / (len(evaluations) - invalid_count),
+        ],
+    }
+    jdump(results, os.path.join(save_path, "battle_results.json"))
+
+    print(f" Total {invalid_count} invalid score pair(s).")
+    print(f" Model {name2} has {better_count} better answer(s).")
+    print(f" Model {name2} has {worse_count} worse answer(s).")
+    print(f" {tie_count} answer(s) play(s) to a tie.")
+    print(f" Win rate of model {name2}: {better_count/(len(evaluations)-invalid_count):.2f}")
+    print(f" Model {name1} average score: {ans1_score/(len(evaluations)-invalid_count):.2f}")
+    print(f" Model {name2} average score: {ans2_score/(len(evaluations)-invalid_count):.2f}")
+
+
+def get_gpt35_evaluation(prompt: Dict[str, Any], inst: Dict[str, Any], max_tokens: int = 2048) -> Dict[str, Any]:
+    """
+    Use GPT-3.5 to evaluate one model answer.
+
+    Args:
+        prompt: a dictionary including prompt template, CoT and metrics.
+        inst: the instruction that is needed to be evaluated.
+        max_tokens: the maximum number of tokens to generate in the completion.
+
+    Returns:
+        An evaluation of one answer.
+    """
+
+    MAX_API_RETRY = 3
+
+    question = (inst["instruction"] if inst["input"] == "" else inst["instruction"] + " " + inst["input"])
+    answer = inst["output"]
+    metrics = prompt["metrics"]
+    inst["evaluation"] = {}
+
+    for metric in metrics.keys():
+        for i in range(MAX_API_RETRY):
+            try:
+                response = openai.Completion.create(
+                    model="text-davinci-003",
+                    prompt=prompt["prompt"].format(
+                        question=question,
+                        answer=answer,
+                        metric=prompt["metrics"][metric],
+                        steps=prompt["CoT"][metric],
+                    ),
+                    logprobs=5,
+                    temperature=0,
+                    max_tokens=max_tokens,
+                )
+                inst["evaluation"][metric] = {
+                    "response": response["choices"][0]["text"],
+                    "logprobs": response["choices"][0]["logprobs"]["top_logprobs"],
+                }
+                break
+            except Exception as e:
+                print(e)
+                time.sleep(1)
+    return inst
+
+
+def gpt35_evaluate(
+    answers: List[Dict],
+    prompts: List[Dict],
+    model_name: str,
+    ignore_categories: List[str],
```

Comment:
We do not need `ignore_categories`.

Everything will be specified in the config file, which means that we can also use GPT-3.5 to evaluate `classification` if we specified.